pkg/apis/experimental/types.go

```diff
@@ -639,3 +639,36 @@ type ClusterAutoscalerList struct {
 
 	Items []ClusterAutoscaler `json:"items"`
 }
+
+// Selector is a label query over a set of API objects. MatchLabels and
+// MatchExpressions are ANDed. An empty Selector matches all objects. A
+// null Selector matches no objects.
+type Selector struct {
+	// MatchLabels is a list of labels.
+	MatchLabels map[string]string `json:"matchLabels,omitempty"`
+	// MatchExpressions is a list of requirement selectors.
+	MatchExpressions []SelectorRequirement `json:"matchExpressions,omitempty"`
+}
+
+// SelectorRequirement is a selector that contains values, a key
+// and an operator that relates the key and values.
+type SelectorRequirement struct {
+	// Key is the label key that the selector applies to.
+	Key string `json:"key"`
+	// Operator represents a key's relationship
+	// to a set of values.
+	Operator SelectorOperator `json:"operator"`
+	// StringValues is a set of string values. If the operator is In or NotIn,
+	// the values set must be non-empty.
+	StringValues []string `json:"stringValues,omitempty"`
+}
+
+// SelectorOperator is the set of operators that can be used in
+// a selector requirement.
+type SelectorOperator string
+
+const (
+	SelectorOpIn     SelectorOperator = "In"
+	SelectorOpNotIn  SelectorOperator = "NotIn"
+	SelectorOpExists SelectorOperator = "Exists"
```

Comment:
What are the semantics of "Exists"? Does it mean "require that this key exists, ignore the value"? If so, that's reasonable (and we should have that) -- but shouldn't we also have "DoesNotExist" ("require that this key does not exist")? 

Also, I'd strongly suggest that we have the standard numeric comparison operators. (Where the first value of the StringValues array is interpreted as a floating point number, and the rest of the array is ignored) I think there are a lot of potential use cases for this when the selector is being used as a nodeSelector for a pod, especially on bare metal environments where the cluster admin may not want to hide hardware specs from the user. It's true that in many cases you could just represent that as a new resource, but that requires more work than just adding a label to the node. Also, I think there are scenarios we might not foresee now, were numeric comparison would be useful.